sw/device/tests/sim_dv/rv_dm_ndm_reset_req.c

```diff
@@ -0,0 +1,264 @@
+
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_adc_ctrl.h"
+#include "sw/device/lib/dif/dif_flash_ctrl.h"
+#include "sw/device/lib/dif/dif_keymgr.h"
+#include "sw/device/lib/dif/dif_otp_ctrl.h"
+#include "sw/device/lib/dif/dif_pinmux.h"
+#include "sw/device/lib/dif/dif_rstmgr.h"
+#include "sw/device/lib/dif/dif_sysrst_ctrl.h"
+#include "sw/device/lib/testing/rstmgr_testutils.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#include "adc_ctrl_regs.h"
+#include "flash_ctrl_regs.h"
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "keymgr_regs.h"
+#include "otp_ctrl_regs.h"
+#include "pinmux_regs.h"
+#include "sysrst_ctrl_regs.h"
+/*
+   RV_DM NDM RESET REQUEST TEST
+   In top_earlgrey, csr can be devided 3 groups as below.
+     1. Group1 : Devce under por_reset
+        - pwrmgr, rstmgr
+     2. Group2 : Device under lc_reset
+        - otp_ctrl, pinmux, rv_dm and aontimer
+
+       * clkmgr stay between category 1 and 2.
+     3. Group3 : None of the above
+
+   Upon power up, test programs following registers from Group 2 and 3.
+     Group2:
+                                          RESET    PRGM (ARBITRARY VALUE)
+       OTP_CTRL.DIRECT_ACCESS_WDATA0      0x0      0x0609_2022
+       PINMUX.WKUP_DETECTOR_CNT_TH_1      0X0      0X44 --> move to LC
+       SRAM RET ADDRESS(8)                ?        0xDDAA_55BB
+    Group3:
+                                          RESET    PRGM (ARBITRARY VALUE)
+       ADC_CTRL.ADC_SAMPLE_CTL            0x9B     0x37
+       SYSRST_CTRL.EC_RST_CTL             0x7D0    0x567
+       KEYMGR.MAX_OWNER_KEY_VER_SHADOWED  0x0      0x1600_ABBA
+       FLASH_CTRL.SCRATCH                 0x0      0x3927
+
+   After programming csrs, the test assert NDM reset from RV_DM and de-assert.
+   Read programmed csr to check all Group2 keep programmed value while group 3
+   csrs have reset values.
+
+ */
+
+/**
+ * Test csr struct
+ */
+typedef struct test_register {
+  /**
+   * Name of the device.
+   */
+  const char *name;
+  /**
+   * Handle to the DIF object for this device.
+   */
+  void *dif_handle;
+  /**
+   * Arbitrary write value to check register reset.
+   * If the register got reset, then this value will get wiped out.
+   */
+  uint32_t write_val;
+  /**
+   * Expected read value.
+   * For Group 2 registers, it is programmed value - arbitrary, hard coded.
+   * For Group 3 registers, it is reset value.
+   * in the beginning of the test.
+   */
+  uint32_t exp_read_val;
+
+} test_register_t;
+
+static dif_rstmgr_t rstmgr;
+static dif_otp_ctrl_t otp_ctrl;
+static dif_flash_ctrl_t flash_ctrl;
+static dif_adc_ctrl_t adc_ctrl;
+static dif_sysrst_ctrl_t sysrst_ctrl;
+static dif_pinmux_t pinmux;
+static dif_keymgr_t keymgr;
+
+enum {
+  OTP_CTRL,
+  PINMUX,
+  SRAM_CTRL,
+  ADC_CTRL,
+  SYSRST_CTRL,
+  KEYMGR,
+  FLASH_CTRL,
+};
+
+static test_register_t kReg[] = {
```

Comment:
let me know if i understood it correctly, from what i remember, i thought the `exp_rd_val` was supposed to be the value after an ndm reset, if so, a couple below look like they might be wrong.